File: aoc2015/day1/a.py
import intcode as ic


def find_floor(parens):
    floor = 0
    basement = None
    for i, p in enumerate(parens):
        if p == '(':
            floor += 1
        elif p == ')':
            floor -= 1
        else:
            print("What?")

        if floor == -1 and basement is None:
            basement = i + 1
    return floor, basement


tester = ic.Tester('lisp')

tester.test_value(find_floor("(())"), (0, None))
tester.test_value(find_floor("()()"), (0, None))
tester.test_value(find_floor("((("), (3, None))
tester.test_value(find_floor("(()(()("), (3, None))
tester.test_value(find_floor('))((((())))'), (-1, 1))
tester.test_value(find_floor(')))'), (-3, 1))
tester.test_value(find_floor(')())())'), (-3, 1))

with open("input") as f:
    data = f.read()

floor, basement = find_floor(data)
tester.test_value(floor, 74)
tester.test_value(basement, 1795)
print(f'Santa ends up on floor {floor} enters the basement at {basement}')

tester.summary()

File: aoc2015/day1/a.c
#include "aoc.h"

typedef struct {
    ssize_t floor;
    ssize_t basement;
} Floor;

static Floor find_floor(char * parens) {
    Floor result = (Floor) {0, -1};

    size_t i = 0;

    while (parens[i] != '\0') {
        char p = parens[i];
        if (p == '(') {
            result.floor++;
        } else if(p == ')') {
            result.floor--;
        } else {
            printf("What?\n");
        }

        i++;
        
        if (result.floor == -1 && result.basement == -1) {
            result.basement = i;
        }
    }

    return result;
}

static bool cmp(Floor f1, Floor f2) {
    return f1.floor == f2.floor && (f1.basement == f2.basement);
}

int main() {
    Tester tester = create_tester("Not quite LISP");

    test(&tester, cmp(find_floor("(())"), (Floor) {0, -1}), "");
    test(&tester, cmp(find_floor("()()"), (Floor) {0, -1}), "");
    test(&tester, cmp(find_floor("((("), (Floor) {3, -1}), "");
    test(&tester, cmp(find_floor("(()(()("), (Floor) {3, -1}), "");
    test(&tester, cmp(find_floor("))((((())))"), (Floor) {-1, 1}), "");
    test(&tester, cmp(find_floor(")))"), (Floor) {-3, 1}), "");
    test(&tester, cmp(find_floor(")())())"), (Floor) {-3, 1}), "");

    char * parens = read_input("../aoc2015/day1/input");
    Floor result = find_floor(parens);
    ssize_t floor = result.floor;
    ssize_t basement = result.basement;
    test(&tester, floor == 74, "");
    test(&tester, basement == 1795, "");

    printf("Santa ends up on floor %zd enters the basement at %zd\n", floor, basement);

    test_summary(&tester);
}
